server/routerlicious/packages/lambdas/src/alfred/index.ts

```diff
@@ -621,6 +624,98 @@ export function configureWebSocketServices(
 			};
 		}
 
+		async function disconnectDocument() {
+			// Clear token expiration timer on disconnection
+			clearExpirationTimer();
+			const removeAndStoreP: Promise<void>[] = [];
+			// Send notification messages for all client IDs in the connection map
+			for (const [clientId, connection] of connectionsMap) {
+				if (clientIdConnectionsDisconnected.has(clientId)) {
+					// We already removed this clientId once. Skip it.
+					continue;
+				}
+				const messageMetaData = getMessageMetadata(
+					connection.documentId,
+					connection.tenantId,
+				);
+				logger.info(`Disconnect of ${clientId}`, { messageMetaData });
+				Lumberjack.info(
+					`Disconnect of ${clientId}`,
+					getLumberBaseProperties(connection.documentId, connection.tenantId),
+				);
+
+				// Keep track of disconnected clientIds so that we don't repeat the disconnect signal
+				// for the same clientId if retrying when connectDocument completes after disconnectDocument.
+				clientIdConnectionsDisconnected.add(clientId);
+				connection.disconnect().catch((error) => {
+					const errorMsg = `Failed to disconnect client ${clientId} from orderer connection.`;
+					// Reset disconnection status so we can retry if handling race-condition.
+					clientIdConnectionsDisconnected.delete(clientId);
```

Comment:
should we also reset disconnectP to undefined in catch blocks?

Reply:
In the "disconnect" catch, we specifically want disconnectP to still be defined after failure, otherwise "connect" won't know that disconnect happened already. For "connect" catch, if "disconnect" already happened, there is no way for us to come here again. I think it's unnecessary to reset.